### PR TITLE
[3006.x] Allow exclusion of cmd opts for requisites that run commands, eg. onlyif and unless

### LINKED ISSUE #63956: [BUG] onlyif requisite not working when used in user.present with shell=/sbin/nologin
**Description**
I have observed an oddly specific bug. The [onlyif](https://docs.saltproject.io/en/latest/ref/states/requisites.html#onlyif) requisite fails to work properly, specifically when using it with [user.present](https://docs.saltproject.io/en/latest/ref/states/all/salt.states.user.html#salt.states.user.present) combined with the `shell` argument set to `/sbin/nologin`.

**Setup**

Install salt-master / salt-minion

**Steps to Reproduce the behavior**

_test_onlyif.sls_

```
bob_nologin:
  user.present:
    - name: bob
    - shell: /sbin/nologin
    - onlyif:
      - echo hello world

jim_bash:
  user.present:
    - name: jim
    - shell: /bin/bash
    - onlyif:
      - echo hello world

larry_no_login:
  user.present:
    - name: larry
    - shell: /sbin/nologin
```

Note: The issue in question is observed in the first state id (`bob_no_login`). The other state ids were included in this example to demonstrate that this issue is _not_ present when `onlyif` is present but the `shell` is set to `/bin/bash` or when `shell` is set to `/sbin/nologin` and `onlyif` is not present.

Result:

```
# salt test_minion state.sls test_onlyif
test_minion:
----------
          ID: bob_nologin
    Function: user.present
        Name: bob
      Result: True
     Comment: onlyif condition is false
     Started: 20:42:36.824577
    Duration: 1790.267 ms
     Changes:   
----------
          ID: jim_bash
    Function: user.present
        Name: jim
      Result: True
     Comment: New user jim created
     Started: 20:42:38.615449
    Duration: 74.605 ms
     Changes:   
              ----------
              fullname:
              gid:
                  10012
              groups:
                  - jim
              home:
                  /home/jim
              homephone:
              name:
                  jim
              other:
              passwd:
                  x
              roomnumber:
              shell:
                  /bin/bash
              uid:
                  10012
              workphone:
----------
          ID: larry_no_login
    Function: user.present
        Name: larry
      Result: True
     Comment: New user larry created
     Started: 20:42:38.690626
    Duration: 61.091 ms
     Changes:   
              ----------
              fullname:
              gid:
                  10013
              groups:
                  - larry
              home:
                  /home/larry
              homephone:
              name:
                  larry
              other:
              passwd:
                  x
              roomnumber:
              shell:
                  /sbin/nologin
              uid:
                  10013
              workphone:

Summary for test_minion
------------
Succeeded: 3 (changed=2)
Failed:    0
------------
Total states run:     3
Total run time:   1.926 s
```

**Expected behavior**

The `onlyif` requisite should be recognized as "true" and the state id should execute even when `shell` is set to `/sbin/nologin` (for `bob_nologin` in the above example) 

**Versions Report**

```
Salt Version:
          Salt: 3005.1
 
Dependency Versions:
          cffi: 1.14.6
      cherrypy: unknown
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.0
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.9.8
        pygit2: Not Installed
        Python: 3.9.16 (main, Jan  6 2023, 22:50:07)
  python-gnupg: 0.4.8
        PyYAML: 5.4.1
         PyZMQ: 23.2.0
         smmap: Not Installed
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.4
 
System Versions:
          dist: ubuntu 22.04 jammy
        locale: utf-8
       machine: x86_64
       release: 5.15.0-1031-aws
        system: Linux
       version: Ubuntu 22.04 jammy
```

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Initial pass at adding the ability to exclude arguments from a state that end up passed to cmd.retcode when onlyif or other requisites that call cmd.remote are used.

### What issues does this PR fix or reference?
Fixes: #63956 

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by whytewolf — DISMISSED** [review]

**Review by twangboy — DISMISSED** [review]

File: doc/ref/states/requisites.rst [hunk 1]
```diff
@@ -917,6 +917,29 @@ In the above case, ``some_check`` will be run prior to _each_ name -- once for
               key:  not-existing
               get_return: res
 
+.. versionchanged:: 3006
```
Comment (Ch3LL) [inline doc/ref/states/requisites.rst:920]:
```suggestion
.. versionchanged:: 3006.0
```

File: doc/ref/states/requisites.rst [hunk 2]
```diff
@@ -917,6 +917,29 @@ In the above case, ``some_check`` will be run prior to _each_ name -- once for
...
+
+    Since the ``unless`` requisite utilizes ``cmd.retcode``, certain parameters
+    included in the state are passed along to ``cmd.retcode``.  On occasion this
+    can cause issues, particularly if the ``shell`` option in a ``user.present``
+    is set to /sbin/nologin and this shell is passed along to ``cmd.retcode``.
+    This would cause ``cmd.retcode`` to run the command using that shell which
+    would fail regardless of the result of the command.
+
+    By include ``shell`` in ``cmd_opts_exclude``, that parameter would not be
```
Comment (Ch3LL) [inline doc/ref/states/requisites.rst:929]:
```suggestion
    By including ``shell`` in ``cmd_opts_exclude``, that parameter would not be
```

File: doc/ref/states/requisites.rst [hunk 3]
```diff
@@ -1005,6 +1028,28 @@ if the gluster commands return a 0 ret value.
               key:  does-exist
               get_return: res
 
+.. versionchanged:: 3006
```
Comment (Ch3LL) [inline doc/ref/states/requisites.rst:1031]:
```suggestion
.. versionchanged:: 3006.0
```

File: doc/ref/states/requisites.rst [hunk 4]
```diff
@@ -1005,6 +1028,28 @@ if the gluster commands return a 0 ret value.
...
+
+    Since the ``onlyif`` requisite utilizes ``cmd.retcode``, certain parameters
+    included in the state are passed along to ``cmd.retcode``.  On occasion this
+    can cause issues, particularly if the ``shell`` option in a ``user.present``
+    is set to /sbin/nologin and this shell is passed along to ``cmd.retcode``.
+    This would cause ``cmd.retcode`` to run the command using that shell which
+    would fail regardless of the result of the command.
+
+    By include ``shell`` in ``cmd_opts_exclude``, that parameter would not be
```
Comment (Ch3LL) [inline doc/ref/states/requisites.rst:1040]:
```suggestion
    By including ``shell`` in ``cmd_opts_exclude``, that parameter would not be
```

File: changelog/63956.fixed.md [hunk 5]
```diff
@@ -0,0 +1 @@
+Adding the ability to exclude arguments from a state that end up passed to cmd.retcode when requisites such as onlyif or unless are used.
```
Comment (Ch3LL) [inline changelog/63956.fixed.md:1]: is this a feature? should the extension be added instad of fixed?

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Review by Ch3LL — APPROVED** [review]

**Review by twangboy — APPROVED** [review]

**Pull request merged**